src/details/ArborX_DetailsTreeTraversal.hpp

```diff
@@ -97,41 +100,58 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
 
     Node const *stack[64];
     Node const **stack_ptr = stack;
-    *stack_ptr++ = nullptr;
-    Node const *node = _bvh.getRoot();
+    *stack_ptr++ = nullptr;            // push sentinel value
+    Node const *node = _bvh.getRoot(); // start with root
     do
     {
-      Node const *child_left = _bvh.getNodePtr(node->left_child);
-      Node const *child_right = _bvh.getNodePtr(node->right_child);
+      int left_child_index = node->left_child;
+      int right_child_index = node->right_child;
+
+      Node const *left_child_node = _bvh.getNodePtr(left_child_index);
+      Node const *right_child_node = _bvh.getNodePtr(right_child_index);
+
+      bool left_child_is_leaf = left_child_node->isLeaf();
+      bool right_child_is_leaf = right_child_node->isLeaf();
+
+      bool prune_left =
+          (left_child_is_leaf && pruneLeaf(queryIndex, left_child_index)) ||
+          (!left_child_is_leaf &&
+           pruneLeftSubtree(queryIndex, left_child_index));
+      bool prune_right =
+          (right_child_is_leaf && pruneLeaf(queryIndex, right_child_index));
 
-      bool overlap_left = predicate(_bvh.getBoundingVolume(child_left));
-      bool overlap_right = predicate(_bvh.getBoundingVolume(child_right));
+      bool overlap_left =
+          !prune_left && predicate(_bvh.getBoundingVolume(left_child_node));
+      bool overlap_right =
+          !prune_right && predicate(_bvh.getBoundingVolume(right_child_node));
```

Comment:
I think this reduces readability and debuggability, and would not include prune inside overlap. Instead, I would do 
```c++
bool traverse_left = false;
if (!prune_left)
{
   bool overlap_left = predicate(_bvh.getBoundingVolume(child_left));
   ...
   traverse_left = (overlap_left && !child_left->isLeaf());
}
```

Reply:
I don't understand you. For me, inclusion of prune boolean into overlap clouds the meaning of the overlap. Essentially, it is not longer clear later in the code whether the value of overlap is false because there is no physical overlap, or if is false because it is pruned. For me, overlap only concerns the physical aspect of overlapping two geometric objects, not the overlapping with a tree.

Reply:
The point is not to evaluate the predicate when we determine we want to prune (or "skip" not sure what is the best name for it).
I fail to see how the code you suggest gives you more information than the one I wrote.